src/IROperator.cpp

```diff
@@ -1092,69 +1095,165 @@ Expr widening_sub(Expr a, Expr b) {
+    }
+}
+
 Expr saturating_add(Expr a, Expr b) {
+    user_assert(a.defined() && b.defined()) << "saturating_add of undefined Expr\n";
     match_types(a, b);
     Type result_type = a.type();
     return Call::make(result_type, Call::saturating_add, {std::move(a), std::move(b)}, Call::PureIntrinsic);
 }
 
 Expr saturating_sub(Expr a, Expr b) {
+    user_assert(a.defined() && b.defined()) << "saturating_sub of undefined Expr\n";
     match_types(a, b);
     Type result_type = a.type();
     return Call::make(result_type, Call::saturating_sub, {std::move(a), std::move(b)}, Call::PureIntrinsic);
 }
 
 Expr halving_add(Expr a, Expr b) {
+    user_assert(a.defined() && b.defined()) << "halving_add of undefined Expr\n";
     match_types(a, b);
     Type result_type = a.type();
     return Call::make(result_type, Call::halving_add, {std::move(a), std::move(b)}, Call::PureIntrinsic);
 }
 
 Expr rounding_halving_add(Expr a, Expr b) {
+    user_assert(a.defined() && b.defined()) << "rounding_halving_add of undefined Expr\n";
     match_types(a, b);
     Type result_type = a.type();
     return Call::make(result_type, Call::rounding_halving_add, {std::move(a), std::move(b)}, Call::PureIntrinsic);
 }
 
 Expr halving_sub(Expr a, Expr b) {
+    user_assert(a.defined() && b.defined()) << "halving_sub of undefined Expr\n";
     match_types(a, b);
     Type result_type = a.type();
     return Call::make(result_type, Call::halving_sub, {std::move(a), std::move(b)}, Call::PureIntrinsic);
 }
 
 Expr rounding_halving_sub(Expr a, Expr b) {
+    user_assert(a.defined() && b.defined()) << "rounding_halving_sub of undefined Expr\n";
     match_types(a, b);
     Type result_type = a.type();
     return Call::make(result_type, Call::rounding_halving_sub, {std::move(a), std::move(b)}, Call::PureIntrinsic);
 }
 
+Expr mul_shift_right(Expr a, Expr b, Expr q) {
+    user_assert(a.defined() && b.defined() && q.defined()) << "mul_shift_right of undefined Expr\n";
+    user_assert(q.type().is_uint()) << "mul_shift_right shift must be unsigned\n";
+    match_types(a, b);
+    match_lanes(a, q);
+    match_bits(a, q);
+    // q might have widened the bits or lanes of a.
+    match_types(a, b);
+    Type result_type = a.type();
+    return Call::make(result_type, Call::mul_shift_right, {std::move(a), std::move(b), std::move(q)}, Call::PureIntrinsic);
+}
+
+Expr mul_shift_right(Expr a, Expr b, int q) {
+    user_assert(a.defined() && b.defined()) << "mul_shift_right of undefined Expr\n";
+    Type qt = a.type().with_code(halide_type_uint);
+    check_representable(qt, q);
+    return mul_shift_right(std::move(a), std::move(b), make_const(qt, q));
+}
+
+Expr rounding_mul_shift_right(Expr a, Expr b, Expr q) {
+    user_assert(a.defined() && b.defined() && q.defined()) << "rounding_mul_shift_right of undefined Expr\n";
+    user_assert(q.type().is_uint()) << "rounding_mul_shift_right shift must be unsigned\n";
+    match_types(a, b);
+    match_lanes(a, q);
+    match_bits(a, q);
```

Comment:
I see a lot of logic throughout this PR to handle q being a weird type. I'm not sure when or why this would happen though. If it's an unknown Int(32) scalar param? Is this sort of thing tested? I think I only see constant q below.

Reply:
Ah, I'd forgotten how involved the shifts had gotten. If this just matches the shift logic, then fine. But is the non-constant case tested somewhere?

Reply:
There's testing of (rounding_)lower_mul_shift_right in test/correctness/intrinsics.cpp, which is the same codepath used by codegen for the variable case. There's also tons of usage of the variable shift case in #5891, which I forked this from, and hopefully will be merged soon.